agent/consul/state/catalog.go

```diff
@@ -2104,6 +2108,52 @@ func (s *Store) serviceDumpKindTxn(tx *memdb.Txn, ws memdb.WatchSet, kind struct
 	return s.parseCheckServiceNodes(tx, nil, idx, "", results, err)
 }
 
+func (s *Store) IngressGatewaysForService(ws memdb.WatchSet, serviceName string, entMeta *structs.EnterpriseMeta) (uint64, structs.CheckServiceNodes, error) {
+	tx := s.db.Txn(false)
+	defer tx.Abort()
+
+	_, entries, err := s.configEntriesByKindTxn(tx, ws, structs.IngressGateway, entMeta)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var ingressServices []string
+	for _, entry := range entries {
+		ingressConfig := entry.(*structs.IngressGatewayConfigEntry)
+	ENTRIES_LOOP:
+		for _, listener := range ingressConfig.Listeners {
+			for _, service := range listener.Services {
+				// If the request specified a different namespace, skip this service.
+				namespace := entMeta.NamespaceOrDefault()
+				if namespace != service.NamespaceOrDefault() {
+					continue
+				}
+
+				if service.Name == serviceName || service.Name == structs.WildcardSpecifier {
+					ingressServices = append(ingressServices, ingressConfig.Name)
+					break ENTRIES_LOOP
+				}
+			}
+		}
+	}
+
+	var maxIdx uint64
```

Comment:
should this take into account the indexes of the config entries read above?

Reply:
Should be updated now.

Reply:
This function got ripped out entirely converting over to the memdb mapping table.